File: Python/SistemaLibreria/Libreria.py
class Libreria:

    __listaUsuarios = []
    __listaLibros = []

    def __init__(self):
        self.__listaUsuarios = []
        self.__listaLibros = []
    
    def getListaUsuarios(self):
        return self.__listaUsuarios
    
    def setListaUsuarios(self, listaUsuarios):
        self.__listaUsuarios = listaUsuarios
    
    def getListaLibros(self):
        return self.__listaLibros
    
    def setListaLibros(self, listaLibros):
        self.__listaLibros = listaLibros
    
    def agregarUsuario(self, nuevoUsuario):
        self.__listaUsuarios.append(nuevoUsuario)
    
    def  agregarLibros(self, nuevoLibro):
        self.__listaLibros.append(nuevoLibro)
    
    def mostrarUsuarios(self):
        if not self.__listaUsuarios:
            print("\n\t***No hay usuarios registrados***")
        else:
            print("\n\t***USUARIOS REGISTRADOS***")
            for nuevoUsuario in self.__listaUsuarios:
                print("\nNombre: " + str(nuevoUsuario.getName()) + "\nEdad: " + str(nuevoUsuario.getAge()) + "\nId generado: " + str(nuevoUsuario.getId()))
    
    def mostrarLibros(self):
        if not self.__listaLibros:
            print("\n\tNo hay libros registrados")
        else:
            print("\n\t***LIBROS REGISTRADOS***")
            for nuevoLibro in self.__listaLibros:
                print("\nTítulo: " + str(nuevoLibro.getTittle()) + "\nAutor: " + str(nuevoLibro.getAutor()) + "\nId generado: " + str(nuevoLibro.getId()))
            
    
    def rentarLibro(self, nuevoUsuario, nuevoLibro):
        if not nuevoLibro.getRentado():
            nuevoLibro.rentar()
            nuevoUsuario.rentarLibro(nuevoLibro)
            print("\nEl libro - " + str(nuevoLibro.getTittle()) + " - ha sido rentado por " + str(nuevoUsuario.getName()))
        else:
            print("\nEl libro - " + nuevoLibro.getTittle() + " - no está disponible, ya ha sido rentado.")

    def mostrarUsuariosConLibros(self):
        print("\n***Lista de usuarios que han rentado un libro***")
        for nuevoUsuario in self.__listaUsuarios:
            tieneLibrosRentados = any(libro.getRentado() for libro in nuevoUsuario.getLibrosRentados())
            if tieneLibrosRentados:
                print(str(nuevoUsuario.getName()))

    def mostrarLibrosNoRentados(self):
        print("\n***Lista de libros que no han sido rentados***")
        for nuevoLibro in self.__listaLibros:
            if not nuevoLibro.getRentado():
                print("\nTítulo: " + str(nuevoLibro.getTittle()) + "\nAutor: " + str(nuevoLibro.getAutor()))

    def mostrarLibrosRentados(self):
        print("\n***Lista de libros que han sido rentados***")
        for nuevoLibro in self.__listaLibros:
            if nuevoLibro.getRentado():
                print("\nTítulo: " + str(nuevoLibro.getTittle()) + "\nAutor: " + str(nuevoLibro.getAutor()))

        

File: Java/SistemaLibreria/src/sistemalibreria/Libreria.java
package sistemalibreria;

import java.util.ArrayList;
import java.util.List;

public class Libreria 
{

    private List<Usuarios> listaUsuarios;
    private List<Libro> listaLibros;

    public Libreria()
    {
        this.listaUsuarios = new ArrayList<>();
        this.listaLibros = new ArrayList<>();
    }

    public List<Usuarios> getListaUsuarios() 
    {
        return listaUsuarios;
    }

    public void setListaUsuarios(List<Usuarios> listaUsuarios) 
    {
        this.listaUsuarios = listaUsuarios;
    }

    public List<Libro> getListaLibros() {
        return listaLibros;
    }

    public void setListaLibros(List<Libro> listaLibros) {
        this.listaLibros = listaLibros;
    }
    
    
    public void agregarUsuario(Usuarios nuevoUsuario) {
        listaUsuarios.add(nuevoUsuario);
    }

    public void agregarLibros(Libro nuevoLibro) {
        listaLibros.add(nuevoLibro);
    }

    //Funciones
    public void mostrarUsuarios() 
    {
        if (listaUsuarios.isEmpty()) 
        {
            System.out.println("\n\t***No hay usuarios registrados***");
            return;
        } 
        else 
        {
            System.out.println("\n\t***USUARIOS REGISTRADOS***");
            
            for (Usuarios nuevoUsuario : listaUsuarios) 
            {
                System.out.println("\nNombre: " + nuevoUsuario.getName() + "\nEdad: " + nuevoUsuario.getAge() + "\nId generado: " + nuevoUsuario.getId());
            }
        }

    }
    
    public void mostrarLibros() 
    {
        if (listaLibros.isEmpty()) 
        {
            System.out.println("\n\t***No hay libros registrados***");
            return;
        } 
        else 
        {
            System.out.println("\n\t***LIBROS REGISTRADOS***");
            
            for (Libro nuevoLibro : listaLibros) 
            {
                System.out.println("\nTítulo: " + nuevoLibro.getTittle() + "\nAutor: " + nuevoLibro.getAutor() + "\nId generado: " + nuevoLibro.getId());
            }
        }

    }
    
    public void rentarLibro(Usuarios nuevoUsuario, Libro nuevoLibro) 
    {
    if (!nuevoLibro.isRentado())
    {
        nuevoLibro.rentar();
        nuevoUsuario.rentarLibro(nuevoLibro); // Agregar el libro a los libros rentados por el usuario
        System.out.println("\nEl libro - " + nuevoLibro.getTittle() + " - ha sido rentado por " + nuevoUsuario.getName());
    } 
    else 
    {
        System.out.println("\nEl libro - " + nuevoLibro.getTittle() + " - no está disponible, ya ha sido rentado.");
    }
}
    
     public void mostrarUsuariosConLibros()
     {
    System.out.println("\n***Lista de usuarios que han rentado un libro***");
    for (Usuarios nuevoUsuario : listaUsuarios)
    {
        boolean tieneLibrosRentados = false;
        for (Libro libro : nuevoUsuario.getLibrosRentados()) 
        {
            if (libro.isRentado()) 
            {
                tieneLibrosRentados = true;
                break;
            }
        }
        if (tieneLibrosRentados) 
        {
            System.out.println(nuevoUsuario.getName());
        }
    }
}
    
  
     public void mostrarLibrosNoRentados() {
        System.out.println("\n***Lista de libros que no han sido rentados***");
        for (Libro nuevoLibro : listaLibros) 
        {
            if (!nuevoLibro.isRentado()) 
            {
                System.out.println("\nTítulo: " + nuevoLibro.getTittle() + "\nAutor: " + nuevoLibro.getAutor());
            }
        }
    }

    public void mostrarLibrosRentados()
    {
        System.out.println("\n***Lista de libros que han sido rentados***");
        for (Libro nuevoLibro : listaLibros)
        {
            if (nuevoLibro.isRentado()) 
            {
                System.out.println("\nTítulo: " + nuevoLibro.getTittle() + "\nAutor: " + nuevoLibro.getAutor());
            }
        }
    }

}
